pkg/kubectl/cmd/cp.go

```diff
@@ -0,0 +1,299 @@
+		kubectl cp /tmp/foo <some-pod>:/tmp/bar -c <specific-container>
+
+		# Copy /tmp/foo local file to /tmp/bar in a remote pod in namespace <some-namespace>
+		kubectl cp /tmp/foo <some-namespace>/<some-pod>:/tmp/bar
+
+		# Copy /tmp/foo from a remote pod to /tmp/bar locally
+		kubectl cp <some-namespace>/<some-pod>:/tmp/foo /tmp/bar`)
+
+	cpUsageStr = dedent.Dedent(`
+	    expected 'cp <file-spec-src> <file-spec-dest> [-c container]'.
+	    <file-spec> is:
+		[namespace/]pod-name:/file/path for a remote file
+		/file/path for a local file`)
+)
+
+// NewCmdCp creates a new Copy command.
+func NewCmdCp(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "cp <file-spec-src> <file-spec-dest>",
+		Short:   "Copy files and directories to and from containers.",
+		Long:    "Copy files and directories to and from containers.",
+		Example: cp_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(runCopy(f, cmd, cmdOut, cmdErr, args))
+		},
+	}
+	cmd.Flags().StringP("container", "c", "", "Container name. If omitted, the first container in the pod will be chosen")
+
+	return cmd
+}
+
+type fileSpec struct {
+	PodNamespace string
+	PodName      string
+	File         string
+}
+
+func extractFileSpec(arg string) (fileSpec, error) {
+	pieces := strings.Split(arg, ":")
+	if len(pieces) == 1 {
+		return fileSpec{File: arg}, nil
+	}
+	if len(pieces) != 2 {
+		return fileSpec{}, fmt.Errorf("Unexpected fileSpec: %s, expected [[namespace/]pod:]file/path", arg)
+	}
+	file := pieces[1]
+
+	pieces = strings.Split(pieces[0], "/")
+	if len(pieces) == 1 {
+		return fileSpec{
+			PodName: pieces[0],
+			File:    file,
+		}, nil
+	}
+	if len(pieces) == 2 {
+		return fileSpec{
+			PodNamespace: pieces[0],
+			PodName:      pieces[1],
+			File:         file,
+		}, nil
+	}
+
+	return fileSpec{}, fmt.Errorf("Unexpected file spec: %s, expected [[namespace/]pod:]file/path", arg)
+}
+
+func runCopy(f cmdutil.Factory, cmd *cobra.Command, out, cmderr io.Writer, args []string) error {
+	if len(args) != 2 {
+		return cmdutil.UsageError(cmd, cpUsageStr)
+	}
+	srcSpec, err := extractFileSpec(args[0])
+	if err != nil {
+		return err
+	}
+	destSpec, err := extractFileSpec(args[1])
+	if err != nil {
+		return err
+	}
+	if len(srcSpec.PodName) != 0 {
+		return copyFromPod(f, cmd, out, cmderr, srcSpec, destSpec)
+	}
+	if len(destSpec.PodName) != 0 {
+		return copyToPod(f, cmd, out, cmderr, srcSpec, destSpec)
+	}
+	return cmdutil.UsageError(cmd, "One of src or dest must be a remote file specification")
+}
+
+func copyToPod(f cmdutil.Factory, cmd *cobra.Command, stdout, stderr io.Writer, src, dest fileSpec) error {
+	reader, writer := io.Pipe()
+	go func() {
+		defer writer.Close()
+		err := makeTar(src.File, writer)
+		cmdutil.CheckErr(err)
+	}()
+
+	// TODO: Improve error messages by first testing if 'tar' is present in the container?
+	cmdArr := []string{"tar", "xf", "-"}
```

Comment:
`tar` requirement on the pod should be explicitly specified in the description. Otherwise we'll get users complaining about it, imho.